third_party/format_diff/format_diff.py

```diff
@@ -0,0 +1,155 @@
+      "Either 'yapf' or 'clang-format'. This must be set manually if --binary "
+      "is not 'yapf' or 'clang-format'.")
+  parser.add_argument(
+      "-p",
+      metavar="NUM",
+      default=1,
+      help="Strip the smallest prefix containing P slashes. Set to 0 if "
+      "passing `--no-prefix` to `git diff` or using `svn`")
+
+  # Parse and error-check arguments
+  args, binary_args = parser.parse_known_args()
+  if args.binary not in BINARY_TO_DEFAULT_REGEX:
+    if not args.regex:
+      raise parser.error("If --binary is not 'yapf' or 'clang-format' then "
+                         "--regex must be set.")
+    if not args.lines_style:
+      raise parser.error("If --binary is not 'yapf' or 'clang-format' then "
+                         "--lines-style must be set.")
+  else:
+    # Set defaults based off of --binary.
+    if not args.regex:
+      args.regex = BINARY_TO_DEFAULT_REGEX[args.binary]
+    if not args.lines_style:
+      args.lines_style = args.binary
+
+  if args.lines_style not in ["yapf", "clang-format"]:
+    raise parser.error(f"Unexpected value for --line-style {args.lines_style}")
+
+  return args, binary_args
+
+
+def main():
+  args, binary_args = parse_arguments()
+
+  # Extract changed lines for each file.
+  filename = None
+  lines_by_file = {}
+  for line in sys.stdin:
+    # Match all filenames.
+    match = re.search(r"^\+\+\+\ (.*?/){%s}(\S*)" % args.p, line)
+    if match:
+      filename = match.group(2)
+    if filename is None:
+      continue
+
+    # Match all filenames specified by --regex.
+    if not re.match("^%s$" % args.regex, filename):
+      continue
+
+    # Match unified diff line numbers.
+    match = re.search(r"^@@.*\+(\d+)(,(\d+))?", line)
+    if match:
+      start_line = int(match.group(1))
+      line_count = 1
+      if match.group(3):
+        line_count = int(match.group(3))
+      if line_count == 0:
+        continue
+      end_line = start_line + line_count - 1
+
+      if args.lines_style == "yapf":
+        lines = ["--lines", f"{start_line}-{end_line}"]
+      elif args.lines_style == "clang-format":
+        lines = ["-lines", f"{start_line}:{end_line}"]
+      lines_by_file.setdefault(filename, []).extend(lines)
+
+  # Pass the changed lines to --binary alongside any 'unknown' args (e.g. -i).
+  for filename, lines in lines_by_file.items():
+    command = [args.binary, filename]
+    command.extend(lines)
+    command.extend(binary_args)
+
+    print(f"Running `{' '.join(command)}`")
+    p = subprocess.Popen(command,
+                         stdout=subprocess.PIPE,
+                         stderr=None,
+                         stdin=subprocess.PIPE,
+                         universal_newlines=True)
+    stdout, stderr = p.communicate()
+    if p.returncode != 0:
+      sys.exit(p.returncode)
+
+    # Print --binary's output if in-place formatting isn't specified.
+    if "-i" not in binary_args:
```

Comment:
This makes "-i" effectively a known arg, just with weird parsing

Reply:
Proposed fix in https://github.com/google/iree/pull/3482/commits/b80fabb66b70da455b820fb63d1fd829acbb1a13

Reply:
Works for me. We've already probably put too much thought into this :grin: